src/effects/effectchain.cpp

```diff
@@ -1,374 +1,420 @@
 #include "effects/effectchain.h"
 
-#include "effects/effectchainmanager.h"
-#include "effects/effectprocessor.h"
+#include "control/controlencoder.h"
+#include "control/controlpotmeter.h"
+#include "control/controlpushbutton.h"
+#include "effects/effectslot.h"
 #include "effects/effectsmanager.h"
-#include "effects/effectxmlelements.h"
+#include "effects/effectsmessenger.h"
+#include "effects/presets/effectchainpresetmanager.h"
 #include "engine/effects/engineeffectchain.h"
-#include "engine/effects/engineeffectrack.h"
-#include "engine/effects/message.h"
 #include "engine/engine.h"
+#include "mixer/playermanager.h"
 #include "moc_effectchain.cpp"
 #include "util/defs.h"
+#include "util/math.h"
 #include "util/sample.h"
 #include "util/xml.h"
 
-EffectChain::EffectChain(EffectsManager* pEffectsManager, const QString& id,
-                         EffectChainPointer pPrototype)
-        : QObject(pEffectsManager),
+EffectChain::EffectChain(const QString& group,
+        EffectsManager* pEffectsManager,
+        EffectsMessengerPointer pEffectsMessenger,
+        SignalProcessingStage stage)
+        : // The control group names are 1-indexed while internally everything
+          // is 0-indexed.
+          m_presetName(""),
           m_pEffectsManager(pEffectsManager),
-          m_pPrototype(pPrototype),
-          m_bEnabled(true),
-          m_id(id),
-          m_name(""),
-          m_mixMode(EffectChainMixMode::DrySlashWet),
-          m_dMix(0),
-          m_pEngineEffectChain(nullptr),
-          m_bAddedToEngine(false) {
+          m_pChainPresetManager(pEffectsManager->getChainPresetManager()),
+          m_pMessenger(pEffectsMessenger),
+          m_group(group),
+          m_signalProcessingStage(stage),
+          m_pEngineEffectChain(nullptr) {
+    // qDebug() << "EffectChain::EffectChain " << group << ' ' << iChainNumber;
+
+    m_pControlClear = std::make_unique<ControlPushButton>(ConfigKey(m_group, "clear"));
+    connect(m_pControlClear.get(),
+            &ControlObject::valueChanged,
+            this,
+            &EffectChain::slotControlClear);
+
+    m_pControlNumEffectSlots = std::make_unique<ControlObject>(
+            ConfigKey(m_group, "num_effectslots"));
+    m_pControlNumEffectSlots->setReadOnly();
+
+    m_pControlNumPresetsAvailable = std::make_unique<ControlObject>(
+            ConfigKey(m_group, "num_presetsavailable"));
```

Comment:
This was added by @Holzhaus in https://github.com/Be-ing/mixxx/pull/50 . I don't think it's needed.

Reply:
> I have the impression that this is only clutter and can be removed, because we already have a API for anonymously change effects and fro picking effects by name.

IIRC this was added to make it possible to select a preset using SHIFT+Knob Turn, which is consistent with what we offer for regular FX units in out standard effects mapping.

Reply:
Yes, and I still think this is not a good idea. Selecting discrete values with continuous knobs is not really useful. That should be done with either buttons or a stepped encoder.

Reply:
It works perfectly and is consistent with our standard effects mapping. And most controllers don't have a stepped encoder or buttons available for that, so we shouldn't gatekeep. Without this, there is no way to change the quickeffect preset while DJing and makes the main feature of this PR useless for me.

Reply:
I think adding that to the standard effects mapping was a mistake. Since I implemented it, I have not used it once, neither while mixing nor while developing/testing, because it is not really useful.

Reply:
> Without this, there is no way to change the quickeffect preset while DJing and makes the main feature of this PR useless for me.

That is not true. As I suggested before, you can map `loaded_preset` to buttons.

Reply:
> I think adding that to the standard effects mapping was a mistake. Since I implemented it, I have not used it once, neither while mixing nor while developing/testing, because it is not really useful.

There are loads of features in Mixxx that I have never used. But I guess most of them have their place. And for me, not being able to select a different effect while DJing would be a no-go, so if you hadn't implemented it, I probably would have ;-) 

> That is not true. As I suggested before, you can map loaded_preset to buttons.

I don't have an additional button on my controller, and I'm not planning to buy a new one. And even if I had, I don't think pressing a button a bunch of times is necessarily better than putting a knob in a certain position, at least if I only have a few presets in list of presets.

Also, `loaded_preset` just does nothing if you set it to a value bigger than the number of presets. So you need a way to query how many presets you have anyway.

Reply:
This does not add much complexity to the code so I'm fine leaving it how it is. I don't care much either way. But I'd highly encourage mapping `loaded_preset` to buttons or stepped encoders instead of continuous knobs or faders.